File: src/Algorithms/Lists/Three_Computational_Problems/how_sum.py
def how_sum(target_sum, numbers):
    if target_sum < 0:
        return None
    if target_sum == 0:
        return []

    for num in numbers:
        remainder = target_sum - num
        remainder_result = how_sum(remainder, numbers)
        if (remainder_result is not None):
            remainder_result.append(num)
            return remainder_result

    return None

File: src/Algorithms/Lists/Three_Computational_Problems/how-sum.ts
// O(n^m * m) time
// O(m) space
// Where m = target sum, n = array length
const howSumMemo = (
  targetSum: number,
  numbers: number[]
  //   memo: Record<number, boolean> = {}
): number[] | null => {
  if (targetSum < 0) return null;
  if (targetSum === 0) return [];

  for (let num of numbers) {
    const remainder = targetSum - num;
    const remainderResult = howSumMemo(remainder, numbers);
    if (remainderResult !== null) {
      return [...remainderResult, num];
    }
  }

  return null;
};

export const solutions = {
  howSumMemo,
};
